### PR TITLE
[SPARK-45685][SQL][FOLLOWUP] Add handling for `Stream` where `LazyList.force` is called

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Refer to the suggestion of https://github.com/apache/spark/pull/43563#pullrequestreview-2114900378, this pr add handling for `Stream` where `LazyList.force` is called.

### Why are the changes needed?
Even though `Stream` is deprecated in 2.13, it is not _removed_ and thus is is possible that some parts of Spark / Catalyst (or third-party code) might continue to pass around `Stream` instances. Hence, we should restore the call to `Stream.force` where `.force` is called on `LazyList`, to avoid losing the eager materialization for Streams that happen to flow to these call sites. This is also a guarantee of compatibility.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add some new tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -226,12 +226,14 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
...
 
+    @scala.annotation.nowarn("cat=deprecation")
     def recursiveTransform(arg: Any): AnyRef = arg match {
       case e: Expression => transformExpression(e)
       case Some(value) => Some(recursiveTransform(value))
       case m: Map[_, _] => m
       case d: DataType => d // Avoid unpacking Structs
-      case stream: LazyList[_] => stream.map(recursiveTransform).force
+      case stream: Stream[_] => stream.map(recursiveTransform).force
+      case lazyList: LazyList[_] => lazyList.map(recursiveTransform).force
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:236]: cc @JoshRosen Did I understand your suggestion correctly? Thanks ~

Comment (JoshRosen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:236]: @LuciferYang, yes, this is exactly what I had in mind.

**Review by JoshRosen — APPROVED** [review]
The code changes here look good to me (and are in line with what I suggested in my comment at https://github.com/apache/spark/pull/43563#pullrequestreview-2114900378 reporting the potential issue). Thanks for also updating the unit tests to ensure test coverage of the new branches.

LGTM pending filling out a PR description.

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @JoshRosen ~

**Pull request closed without merging**